test/cpp/server/server_builder_test.cc

```diff
@@ -90,9 +90,13 @@ TEST_F(ServerBuilderTest, CreateServerRepeatedPortWithDisallowedReusePort) {
 
 TEST_F(ServerBuilderTest, AddPassiveListener) {
   std::unique_ptr<experimental::PassiveListener> passive_listener;
-  ServerBuilder()
-      .AddPassiveListener(InsecureServerCredentials(), passive_listener)
-      .BuildAndStart();
+  auto server =
+      ServerBuilder()
+          .AddPassiveListener(InsecureServerCredentials(), passive_listener)
+          .BuildAndStart();
+  // The listener must be released before the server can shut down.
```

Comment:
This should not be necessary.  I think this is a side effect of the fact that you changed `Server` to hold `RefCountrfPtr` instead of `OrphanablePtr`.

Same thing in two places below.

Reply:
I think this limitation would not accomplish the API guarantee that we've been striving for here, which is that the application doesn't need to closely synchronize shutting down the server with its use of `PassiveListener`s.  That's the whole reason that we wanted `PassiveListener` to hold a ref to the `Chttp2ServerListener` in the first place, so if holding that listener is causing shutdown to hang, then we haven't actually solved this problem.

It looks like the problem is that `Chttp2ServerListener` doesn't invoke the `on_destroy_done_` callback until it's destroyed, but the server won't shut down until that callback is invoked.  I think what we need here is a way to have `Chttp2ServerListener` invoke that callback when all refs *except* the one held by `PassiveListener` go away.

I suggest the following approach.  First, in `PassiveListenerImpl`:

```
class PassiveListenerImpl {
 public:
  // To be called by Chttp2ServerListener's dtor.
  void ListenerDestroyed() {
    MutexLock lock(&mu_);
    listener_ = nullptr;
  }

  void AcceptConnectedEndpoint(
      std::unique_ptr<grpc_event_engine::experimental::EventEngine::Endpoint>
          endpoint) {
    RefCountedPtr<Chttp2ServerListener> listener;
    {
      MutexLock lock(&mu_);
      if (listener_ != nullptr) listener = listener_->RefIfNonZero();
    }
    if (listener == nullptr) {
      return absl::UnavailableError("passive listener already shut down");
    }
    ExecCtx exec_ctx;
    listener_->AcceptConnectedEndpoint(std::move(endpoint));
  }

 private:
  Mutex mu_;
  Chttp2ServerListener* listener_ ABSL_GUARDED_BY(&mu_);
};
```

Then in `Chttp2ServerListener`:

```
class Chttp2ServerListener : public Server::ListenerInterface {
 public:
  ~Chttp2ServerListener() override {
    if (passive_listener_ != nullptr) {
      passive_listener_->ListenerDestroyed();
    }
    // ...
  }

 private:
  // For access to RefIfNonZero().
  friend class PassiveListenerImpl;

  std::shared_ptr<PassiveListenerImpl> passive_listener_;
};
```

Another alternative would be to change `ListenerInterface` to use `std::shared_ptr<>` so that `PassiveListener` can hold a `std::weak_ptr<>` to it, thus not blocking destruction.  But I'd prefer the locking approach described above.

Reply:
The inverted ownership works, nice design.